core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+     *
+     * @param allocationId    allocationId of the recovering shard
+     * @param localCheckpoint the local checkpoint of the shard in question
+     */
+    public void markAllocationIdAsInSync(String allocationId, long localCheckpoint) {
+        ensurePrimary();
+        getEngine().seqNoService().markAllocationIdAsInSync(allocationId, localCheckpoint);
+        // make sure the in sync shard gets an up to date global checkpoint in the case that there is no ongoing indexing
+        globalCheckpointSyncer.accept(shardId);
+    }
+
+    public long getLocalCheckpoint() {
+        return getEngine().seqNoService().getLocalCheckpoint();
+    }
+
+    public long getGlobalCheckpoint() {
+        return getEngine().seqNoService().getGlobalCheckpoint();
+    }
+
+    /**
+     * updates the global checkpoint on a replica shard (after it has been updated by the primary).
+     */
+    public void updateGlobalCheckpointOnReplica(long checkPoint) {
+        ensureReplica();
+        getEngine().seqNoService().updateGlobalCheckpointOnReplica(checkPoint);
+    }
+
+    /**
+     * Notifies the service of the current allocation ids in the cluster state.
+     * see {@link GlobalCheckpointService#updateAllocationIdsFromMaster(Set, Set)} for details.
+     *
+     * @param activeAllocationIds       the allocation ids of the currently active shard copies
+     * @param initializingAllocationIds the allocation ids of the currently initializing shard copies
+     */
+    public void updateAllocationIdsFromMaster(Set<String> activeAllocationIds, Set<String> initializingAllocationIds) {
+        ensurePrimary();
+        Engine engine = getEngineOrNull();
+        // if engine is not yet started, we are not ready yet and can just ignore this
+        if (engine != null) {
+            engine.seqNoService().updateAllocationIdsFromMaster(activeAllocationIds, initializingAllocationIds);
+        }
+    }
+
+    abstract class ScheduledTask extends AbstractRunnable {
+
+        final String name;
+        volatile TimeValue interval;
+        volatile ScheduledFuture<?> scheduledFuture;
+
+        public ScheduledTask(String name, TimeValue interval) {
+            this.name = name;
+            this.interval = interval;
+            this.scheduledFuture = null;
+        }
+
+        public void startIfNeeded() {
+            synchronized (mutex) {
+                if (scheduledFuture != null) {
+                    return;
+                }
+                if (shouldRun()) {
+                    logger.debug("scheduling [{}] every [{}]", name, interval);
+                    reschedule();
+                } else {
+                    logger.debug("scheduled [{}] disabled", name);
+                }
+            }
+        }
+
+        /** returns true if this task should be running for this shard and setting */
+        protected boolean shouldRun() {
```

Comment:
Nope. Instead what I'm looking at is the logic in the base `shouldRun` method and I think that as it things are designed currently, that logic should always be used and so I just see it simpler to not allow derived classes to override.